File: subarray_with_0sum.java
//O(n square) time complexity
//O(1) auxilary space
import java.util.*;
import java.io.*;

public class ex{
    public static void main(String args[])
    {
        Scanner s = new Scanner(System.in);
        int n=s.nextInt();
        int[] arr = new int[n];
        for(int i=0;i<n;i++)
        {
            arr[i] = s.nextInt();
        }
        int sum=0;
        for(int st=0;st<n;st++)
        {
            for(int e=st;e<n;e++)
            {
                sum +=arr[e];
                if(sum==0)
                {
                    System.out.print("["+arr[st]+".."+arr[e]+"]");
                }
            }
            sum=0;
        }
    }
}

File: subarray_with_0sum.cpp
//O(n) time complexity
//O(n) aux space 
#include<bits/stdc++.h>
#include<set>
using namespace std;
int main()
{
    int n;
    cin>>n;
    int arr[n];
    for(int i=0;i<n;i++)
    {
        cin>>arr[i];
    }
    set<int> s;
    int flag=0,sum=0;
    for(int i=0;i<n;i++)
    {
        sum += arr[i];
        if(sum==0 || s.find(sum)!=s.end())
        {
            flag=1;
        }
        s.insert(sum);
    }
    if(flag)
    {
        cout<<"true";
    }
    else
    {
        cout<<"false";
    }
    

}